### ISSUE TITLE
Debugging parameters

### ISSUE THREAD

**Issue by w00519772** [freqtrade/freqtrade#10122]
I am fairly new to Freqtrade and Python.  I am a professional C# developer.

I am trying to run Freqtrade - on a Linux PC and in a Linux docker container on a Windows 10 PC.

The results are the same.  The launch.json looks like this:
```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python Debugger: Current File",
            "type": "debugpy",
            "request": "launch",
            "program": "${workspaceFolder}/Freqtrade/main.py",
            "console": "integratedTerminal",
            "args": ["trade","-c","config.json"]
        }
    ]
}
```
When I debug it in VS Code:

![BREAKPOINT](https://github.com/freqtrade/freqtrade/assets/29153414/50e50210-ce1d-4ef6-8f6f-20b2f16ae9be)

Yet the confi.json looks this this (I have commented out the personal parts as blank strings):

```
{
    "max_open_trades": 3,
    "stake_currency": "BTC",
    "stake_amount": 0.05,
    "tradable_balance_ratio": 0.99,
    "fiat_display_currency": "GBP",
    "timeframe": "5m",
    "dry_run": true,
    "dry_run_wallet": 1000,
    "cancel_open_orders_on_exit": false,
    "trading_mode": "spot",
    "margin_mode": "",
    "unfilledtimeout": {
        "entry": 10,
        "exit": 10,
        "exit_timeout_count": 0,
        "unit": "minutes"
    },
    "entry_pricing": {
        "price_side": "same",
        "use_order_book": true,
        "order_book_top": 1,
        "price_last_balance": 0.0,
        "check_depth_of_market": {
            "enabled": false,
            "bids_to_ask_delta": 1
        }
    },
    "exit_pricing":{
        "price_side": "same",
        "use_order_book": true,
        "order_book_top": 1
    },
    "download_trades": true,
    "exchange": {
        "name": "kraken",
        "key": "",
        "secret":",
        "ccxt_config": {},
        "ccxt_async_config": {},
        "pair_whitelist": [
        ],
        "pair_blacklist": [

        ]
    },
    "pairlists": [
        {
            "method": "VolumePairList",
            "number_assets": 20,
            "sort_key": "quoteVolume",
            "min_value": 0,
            "refresh_period": 1800
        }
    ],
    "telegram": {
        "enabled": false,
        "token": "",
        "chat_id": ""
    },
    "api_server": {
        "enabled": true,
        "listen_ip_address": "127.0.0.1",
        "listen_port": 8080,
        "verbosity": "error",
        "enable_openapi": false,
        "jwt_secret_key": "",
        "ws_token": "",
        "CORS_origins": [],
        "username": "freqtrader",
        "password": ""
    },
    "bot_name": "freqtrade",
    "initial_state": "running",
    "force_entry_enable": false,
    "internals": {
        "process_throttle_secs": 5
    }
}
```

Why is dry run false after the config is supplied?

The Terminal looks like this when I debu:

ftuser@e88b21eef8bf:/workspaces/freqtrade$  /usr/bin/env /usr/local/bin/python /home/ftuser/.vscode-server/extensions/ms-python.debugpy-2024.2.0-linux-x64/bundled/libs/debugpy/adapter/../../debugpy/launcher 36405 -- /workspaces/freqtrade/Freqtrade/main.py trade -c config.json

**Comment by xmatthias** [comment]
Well your breakpoint is simply too early.

At this point - no config has been ~harmed~loaded - so what you get is the default parameters for the command line. 
if you start a different command (not trade, but something like `freqtrade webserver`) - you'll notice that the parameters at this point change (and dry-run will be missing as it's not a valid argument for this subcommand).

in case of dry-run - the default is false (you didn't provide this parameter). That's an implicit default, really - due to the fact that we use [argparse's `store_true`](https://docs.python.org/3/library/argparse.html#action):

(from the above link - unfortunately it's not deep-linkable to exactly this spot)
> 'store_true' and 'store_false' - These are special cases of 'store_const' used for storing the values True and False respectively. In addition, they create default values of False and True respectively. For example:


Simply put - debug further until you're after config parsing - in case of trade mode - that'll be once you're into `worker.py` - though worker is only really used by trade mode - it'll provide you with the correct location.

**Comment by xmatthias** [comment]
As a side note:
the [documentation](https://www.freqtrade.io/en/stable/developer/#debug-configuration) has a recommended debug configuration. 
That'll make sure you're using it as a user would.  While it doesn't make a real difference to what you're trying to do now - it might do in the future (using main.py sometimes leads to odd side-effects).

**Comment by w00519772** [comment]
Thanks for your input, @xmatthias, it helped.  I do have a few further questions though:

1) If I use the debug configuration in the documentation (https://www.freqtrade.io/en/stable/developer/#debug-configuration), then it prompts me for a python file to debug because none is specified - I assume you just always select main.py?
2) If I open the VSCode terminal and type DIR, then use a bash line inside the docker container; there appear to be missing files inside the container e.. config.json.  Here is docker: 

![docker](https://github.com/freqtrade/freqtrade/assets/29153414/16a8f717-d94d-4549-b5c6-4125d2052819)

Here is vscode:

![vscode](https://github.com/freqtrade/freqtrade/assets/29153414/894dc683-89c2-4d31-ab80-254b663afa80)

3) In your debug configuration; it refers to a strategy (MyAwesomeStrategy).  Does this strategy exist or is it one I would create myself?

**Comment by xmatthias** [comment]
the "current dir" is not intended to be part of the container - only things inside user_data is.

As for the debugging - the configuration in the documentation uses `module` - so no file is required.
if you're asked for module, then you're either in the wrong virtual environment (not the one freqtrade is installed in) - or you didn't complete the installation steps (`pip install -e .`  is missing - so it's not a module you can run).

That strategy obviously doesn't exist - it's used as a sample - and reading the docs, you'll notice that we use MyAwesomeStrategy in most examples. That's however not an actual existing strategy (at least not as far as i'm aware) - so obviously, you'll want to adjust the args the way that fit your needs.

**Comment by w00519772** [comment]
@xmatthias, I just installed Docker; cloned the Freqtrade repository; opened it in VS Code and ran it.  What file should be the entry point in the default config (https://www.freqtrade.io/en/stable/developer/#debug-configuration)  - a relative file path would help?  I am not sure what you mean by` pip install -e `.  I have looked here: https://www.freqtrade.io/en/0.18.1/installation/#automatic-installation-docker and there is no reference to it.

**Comment by xmatthias** [comment]
your link goes to a freqtrade version that's probably 5 years old - i don't realy care what is written there, as it's definitely no longer relevant.
We switched to "year" based versioning in 2019. Please use a documentation page that fits your version to avoid problems and missunderstandings.

---

you can't really debug "like this" (trying to start a regular program, while attempting to run it inside docker).

if you'd like to debug _in docker_, then please use the devcontainer (vscode should ask you to reopen in the devcontainer, assuming you have the proper extension installed).

**Comment by w00519772** [comment]
I am debugging inside a dev container (when I open the folder in vscode it prompts me to reopen in container which I do and it succeeds).  When I use my own config.json; I am able to start debugging i.e. main.py is reached.  Here is my config.json: 

```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python Debugger: Current File",
            "type": "debugpy",
            "request": "launch",
            "program": "${workspaceFolder}/Freqtrade/main.py",
            "console": "integratedTerminal",
            "args": ["trade","-c","config.json"]
        }
    ]
}
```

When I use the config,json recommended I am prompted for a Python debugger like this:

![DebuggerPrompt](https://github.com/freqtrade/freqtrade/assets/29153414/229ea02b-d6b8-44fa-adb3-38200ca44b34)

Here is the config.json recommended (https://www.freqtrade.io/en/2024.1/developer/):

```
{
    "name": "freqtrade trade",
    "type": "python",
    "request": "launch",
    "module": "freqtrade"
    "console": "integratedTerminal",
    "args": [
        "trade",
        // Optional:
        // "--userdir", "user_data",
        "--strategy", 
        "MyAwesomeStrategy",
    ]
}
```

@xmatthias , any idea why my config.json works and the config recommended doesn't? Thanks for your help.

**Comment by xmatthias** [comment]
well your 2nd config is invalid as it's missing a comma after the "module" line.

fixing that - and the debugging in the workspace works just fine for me.

![image](https://github.com/freqtrade/freqtrade/assets/5024695/fcaa14da-1e04-400b-8f4b-7fd7c69656a3)

Now i've updated the "type" to debugpy in the docs as that's the new way - but "python" will work just fine (though with a yellow squiggly line showing that it's deprecated).

**Comment by w00519772** [comment]
Thanks.  That is now working as expected.  Thanks @xmatthias .

By the way.  I followed the instructions on this webpage https://www.freqtrade.io/en/latest/telegram-usage/ a few months ago; ran the bot and saw the output in Telegram as expected.

The next day I discovered that Telegram had banned my account (telephone number) and it is still banned.  This is all I have used Telegram for.  Has anyone else experienced this? I have emailed Telegram a few times, but they are yet to reply.

**Comment by xmatthias** [comment]
well support for most bigger services is bad.

I've got multiple bots running though - and haven't had any problem yet - and it's also the first time i hear from an account that's actually banned.

**Comment by stash86** [comment]
Never heard anyone got banned from telegram because of Freqtrade
